File: TradeMinder.UnitTests/FullTests.fs
module FullTests
open NUnit.Framework
open StockApi
open Database
open Messaging
open System
open FsUnit

[<Test>]
let ``Full test: when stock is within thresholds, should not create a message.``() =
    // Prepare return values
    let stock = { StockInfo.Symbol = "MSFT"; Date = DateTime.Now; Value = 10.0M }
    let thresholds = { NotificationThresholds.Symbol = "MSFT"; Email = "jmarr@microdesk.com"; High = 15.0M; Low = 4.0M }
    
    // Stub the StockApi
    let getLatest symbol = 
        symbol |> should equal "MSFT" // Assert expected argument value
        async { return Some stock } 

    // Stub the Database
    let getThresholds symbol email =
        symbol |> should equal "MSFT"
        email |> should equal "jmarr@microdesk.com"
        async { return Some thresholds }

    // Stub Messaging
    let sendMessage (msg: Message) =
        async { return () }
        
    // Run
    FullyTestable.StockThresholdNotifier.checkStockTemplate getLatest getThresholds sendMessage "MSFT" "jmarr@microdesk.com"
    |> Async.RunSynchronously

File: TradeMinderCSharp.UnitTests/FullTests.cs
using Microsoft.VisualStudio.TestTools.UnitTesting;
using Moq;
using System;
using System.Threading.Tasks;
using TradeMinderCSharp.Entities;
using TradeMinderCSharp.Interfaces;

namespace TradeMinderCSharp.UnitTests
{
    [TestClass]
    public class FullTests
    {
        [TestMethod]
        public void FullTest_WhenStockIsWithinThresholds_ShouldNotCreateMessage()
        {
            // Prepare return values
            var stock = new StockInfo { Symbol = "MSFT", Date = DateTime.Now, Value = 10.0M };
            var thresholds = new NotificationThresholds { Symbol = "MSFT", Email = "jmarr@microdesk.com", High = 15, Low = 4 };

            // Stub the StockApi
            var stockApi = new Mock<IStockApi>();
            stockApi
                .Setup(s => s.GetStock(It.IsAny<string>()))
                .Returns(Task.FromResult(stock))
                .Verifiable();

            // Stub the Database
            var database = new Mock<IDatabase>();
            database
                .Setup(d => d.GetThresholds(It.IsAny<string>(), It.IsAny<string>()))
                .Returns(Task.FromResult(thresholds));

            // Stub the MessageService
            var messageSvc = new Mock<IMessageService>();
            messageSvc
                .Setup(m => m.SendMessage(It.IsAny<Message>()));

            // Build up feature class
            var notifier = new StockThresholdNotifier(database.Object, stockApi.Object, messageSvc.Object);

            // Run
            notifier.CheckStock("MSFT", "jmarr@microdesk.com")
                .Wait();

            // Assert expected argument value
            stockApi.Verify(s => s.GetStock("MSFT"));

            // Assert expected argument values
            database.Verify(d => d.GetThresholds("MSFT", "jmarr@microdesk.com"));
        }
    }
}
